File: go/storage/inmemory_kv.go
// SPDX-License-Identifier: MIT

package storage

import (
	"context"
	"errors"
	"sync"
)

// InMemoryKeyValueStore is a process-local, volatile IKeyValueStore backed
// by a sync.Mutex-protected map. Mirrors the C# InMemoryKeyValueStore.
//
// Suitable for tests and demos. Loses everything on process exit.
type InMemoryKeyValueStore struct {
	mu      sync.Mutex
	entries map[string][]byte
}

// NewInMemoryKeyValueStore returns an empty in-memory store.
func NewInMemoryKeyValueStore() *InMemoryKeyValueStore {
	return &InMemoryKeyValueStore{
		entries: make(map[string][]byte),
	}
}

// Get returns the bytes stored under key, or (nil, nil) if absent.
// The returned slice is a defensive copy so callers cannot mutate stored bytes.
func (s *InMemoryKeyValueStore) Get(ctx context.Context, key string) ([]byte, error) {
	if key == "" {
		return nil, errors.New("storage: key cannot be empty")
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	v, ok := s.entries[key]
	if !ok {
		return nil, nil
	}
	out := make([]byte, len(v))
	copy(out, v)
	return out, nil
}

// Put inserts or replaces the bytes stored under key. The store keeps a
// defensive copy so the caller can mutate the original buffer safely.
func (s *InMemoryKeyValueStore) Put(ctx context.Context, key string, value []byte) error {
	if key == "" {
		return errors.New("storage: key cannot be empty")
	}
	if value == nil {
		return errors.New("storage: value cannot be nil")
	}
	cp := make([]byte, len(value))
	copy(cp, value)
	s.mu.Lock()
	defer s.mu.Unlock()
	s.entries[key] = cp
	return nil
}

// Remove deletes the entry under key. Returns (true, nil) if a value was
// removed; (false, nil) if there was nothing to remove.
func (s *InMemoryKeyValueStore) Remove(ctx context.Context, key string) (bool, error) {
	if key == "" {
		return false, errors.New("storage: key cannot be empty")
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	if _, ok := s.entries[key]; !ok {
		return false, nil
	}
	delete(s.entries, key)
	return true, nil
}

// Contains returns true if a value exists under key.
func (s *InMemoryKeyValueStore) Contains(ctx context.Context, key string) (bool, error) {
	if key == "" {
		return false, errors.New("storage: key cannot be empty")
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	_, ok := s.entries[key]
	return ok, nil
}

// ListKeys returns every key currently in the store. Order is unspecified.
func (s *InMemoryKeyValueStore) ListKeys(ctx context.Context) ([]string, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	out := make([]string, 0, len(s.entries))
	for k := range s.entries {
		out = append(out, k)
	}
	return out, nil
}

File: python/aethernet/storage/in_memory_kv.py
# SPDX-License-Identifier: MIT

"""Process-local, volatile :class:`KeyValueStore` backed by a dict.

Suitable for tests and demos. Loses everything on process exit. All
mutations are guarded by a single :class:`asyncio.Lock`, so concurrent
writers see a consistent state even though Python's dict is already
thread-safe — the lock is required because we offer ``list_keys`` as an
async iterator and want a stable snapshot semantics rather than a
"keys may appear or disappear mid-iteration" race.
"""

from __future__ import annotations

import asyncio
from typing import AsyncIterator, Dict, Optional

from aethernet.storage.kv import KeyValueStore


class InMemoryKeyValueStore(KeyValueStore):
    """Volatile in-memory KV store.

    All values are stored as defensive copies so the caller cannot mutate
    the stored bytes after writing them. Likewise, every ``get`` returns a
    fresh copy so the caller cannot mutate the in-memory entry through a
    returned reference.
    """

    def __init__(self) -> None:
        self._entries: Dict[str, bytes] = {}
        self._lock: asyncio.Lock = asyncio.Lock()

    async def get(self, key: str) -> Optional[bytes]:
        if not key:
            raise ValueError("key cannot be empty")
        async with self._lock:
            v = self._entries.get(key)
            return bytes(v) if v is not None else None

    async def put(self, key: str, value: bytes) -> None:
        if not key:
            raise ValueError("key cannot be empty")
        if value is None:
            raise ValueError("value cannot be None")
        async with self._lock:
            self._entries[key] = bytes(value)

    async def remove(self, key: str) -> bool:
        if not key:
            raise ValueError("key cannot be empty")
        async with self._lock:
            if key in self._entries:
                del self._entries[key]
                return True
            return False

    async def contains(self, key: str) -> bool:
        if not key:
            raise ValueError("key cannot be empty")
        async with self._lock:
            return key in self._entries

    async def list_keys(self, prefix: Optional[str] = None) -> AsyncIterator[str]:
        # Snapshot the key set under the lock so the iteration is stable
        # even if writers concurrently mutate the dict.
        async with self._lock:
            keys = list(self._entries.keys())
        for k in keys:
            if prefix is None or k.startswith(prefix):
                yield k
